packages/backend-common/src/tokens/ServerTokenManager.ts

```diff
@@ -86,52 +112,88 @@ export class ServerTokenManager implements TokenManager {
         'Key generation is not supported outside of the dev environment',
       );
     }
+
     if (this.privateKeyPromise) {
       return this.privateKeyPromise;
     }
+
     const promise = (async () => {
-      const secret = await generateSecret('HS256');
+      const secret = await generateSecret(TOKEN_ALG);
       const jwk = await exportJWK(secret);
       this.verificationKeys.push(base64url.decode(jwk.k ?? ''));
       this.signingKey = this.verificationKeys[0];
       return;
     })();
 
     try {
-      // If we fail to generate a new key, we need to clear the state so that
-      // the next caller will try to generate another key.
+      this.privateKeyPromise = promise;
       await promise;
     } catch (error) {
-      this.logger.error(`Failed to generate new key, ${error}`);
+      // If we fail to generate a new key, we need to clear the state so that
+      // the next caller will try to generate another key.
+      this.options.logger.error(`Failed to generate new key, ${error}`);
       delete this.privateKeyPromise;
     }
+
     return promise;
   }
 
   async getToken(): Promise<{ token: string }> {
     if (!this.verificationKeys.length) {
       await this.generateKeys();
     }
-    const sub = 'backstage-server';
-    const jwt = await new SignJWT({ alg: 'HS256' })
-      .setProtectedHeader({ alg: 'HS256', sub: sub })
-      .setSubject('backstage-server')
-      .sign(this.signingKey);
-    return { token: jwt };
+
+    if (this.currentTokenPromise) {
+      return this.currentTokenPromise;
+    }
+
+    const result = Promise.resolve().then(async () => {
```

Comment:
Yeah not actually strictly needed - just makes sure that the actual "hot" function in the `then` gets called on the _next_ tick, out of band, after this method has returned. This was to ensure that it would never accidentally overwrite the promise member with undefined _before it was even written in the first place_. But then I refractored to have the `then`/`catch` separate at the bottom anyway, rendering the initial exercise a bit moot. But still, it ensures a potentially quicker return to the caller. 🤷

Reply:
Not quite the same thing. I still prefer this way of doing it because

- it reads well to me at least (VERY clearly defers the action) since it does the assignment to a member below and i wanted it to be crystal clear that it'll then for sure also be set to undefined in the right order upon failure
- it makes sure that any work that _might_ be sync in `sign()` will be happening inside the promise chain, so that exceptions are caught and handled for sure down at the bottom, not bubbling up and out already here
- and for the same reason as above, any heavy work for sure happens out of band

but i think you are right though that a skipped await on sign() might have basically the same effect in practice in this instance